cs244b_project/src/lib.rs

```diff
@@ -356,31 +362,32 @@ impl StreamletInstance {
-                                        info!("Epoch: {}, (Vote) received VOTE, message {} is NOTARIZED, checking for ancestor chain...", epoch, message.nonce);
-                                        let chain_index = self.blockchain_manager.index_of_ancestor_chain(block.clone());
-                                        match chain_index {
-                                            Some(idx) => {
-                                                info!("Epoch: {}, VOTE message {} descends from a notarized chain. Adding to chain...", epoch, message.nonce);
-                                                self.blockchain_manager
-                                            .       add_to_chain(block.clone(), message.clone().get_signatures(), idx);
-                                                // Only broadcast if we haven't signed already
-                                                if self.sign_message(&mut new_message) {
-                                                    info!("Epoch: {}, (Vote) received VOTE, signing and broadcasting message {}...", epoch, new_message.nonce);
-                                                    net_stack.broadcast_message(new_message.serialize());
-                                                }
-                                            }
-                                            None => {
-                                                info!("Epoch: {}, VOTE, message {} does not descend from a notarized chain. Will not sign or broadcast.", epoch, message.nonce);
-                                            }
+                                    if self.should_vote(&message, voted_this_epoch, epoch, &block, &app_interface) {
+                                        let mut new_message = message.clone();
+                                        // Clone of message that we can modify
+                                        let signed = self.sign_message(&mut new_message);
+                                        // Broadcast messages that we haven't signed yet
+                                        // Note: this is an inexact, but reasonable, proxy for echoing
+                                        if signed {
+                                            info!("Epoch {}: VOTED and signed message {}; broadcasting", epoch, message.nonce);
+                                            net_stack.broadcast_message(new_message.serialize());
+                                        }
+                                        // Once we've voted for a transaction, we should never propose it. 
+                                        self.pending_transactions.retain(|x| *x != block.data);
+                                    }
+                                    // If block is notarized and still extends from a longest notarized chain, 
+                                    // then add it to the chain. 
+                                    if self.is_notarized(&message) {
```

Comment:
...regardless of if "should vote" returns true, if it's notarized, add it to the notarization chain